Add Streamlabs Water Monitor

cpopp opened the pull request:
## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8733

## Example entry for `configuration.yaml` (if applicable):
```yaml
streamlabswater:
  api_key: API_KEY
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

cpopp committed:
Add Streamlabs Water Monitor

homeassistant commented:
Hi @cpopp, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#21205).

Once you do that we will be able to review and accept this pull request.

Thanks!

fabaff commented on homeassistant/components/streamlabswater/__init__.py:
```diff
@@ -0,0 +1,103 @@
+    api_key = conf.get(CONF_API_KEY)
+    location_id = conf.get(CONF_LOCATION_ID)
+
+    client = streamlabswater.StreamlabsClient(api_key)
```
You should check that the API key is valid. If not make the component setup fail. The integration can't recover from this.

cpopp replied:
Thanks for this feedback. Feel free to take a look again and let me know anything else you might find.

cpopp committed:
Fail Streamlabswater component setup when given invalid parameters

The Streamlabs Water component is unable to recover if it is given
an invalid API key or location id so this change is to ensure
we validate they are correct during setup and return a failure
if they are not.

cpopp committed:
Prime Streamlabswater component sensors so data is available immediately

The sensors for the component were not causing an immediate load of
data from the API when being set up so there was some lag after
startup before values would show up.  This change does an explicit
update when the sensors are setup to ensure data is viewable
immediately after startup.

cpopp committed:
Switch Streamlabswater logging to use %s for string formatting

cpopp commented:
Let me know if there is anything else I should address. As far as I'm aware I've addressed all feedback up to this point.

frenck commented:
⚠️ I was unable to find a documentation PR (matching this code change) in our [documentation repository](https://github.com/home-assistant/home-assistant.io).

📚Please, update the documentation and open a PR for it (be sure to create a documentation PR against the `next` branch) and link 🔗 this PR and the new documentation PR by mentioning the link to the PR's in both openings posts/descriptions.

🏷 I am adding the `docs-missing` label until this has been resolved.

👍 Thanks  (times a million!)

cpopp commented:
> ⚠️ I was unable to find a documentation PR (matching this code change) in our [documentation repository](https://github.com/home-assistant/home-assistant.io).
> 
> 📚Please, update the documentation and open a PR for it (be sure to create a documentation PR against the `next` branch) and link 🔗 this PR and the new documentation PR by mentioning the link to the PR's in both openings posts/descriptions.
> 
> 🏷 I am adding the `docs-missing` label until this has been resolved.
> 
> 👍 Thanks (times a million!)

Will do. I had assumed it was fine to just start with the docstrings based on the [developer docs](https://developers.home-assistant.io/docs/en/development_checklist.html), but it sounds like it's not, so when I get some time I'll figure out the docs and open a PR there.

cpopp committed:
Update Streamlabswater component with correct dependencies

Dependencies were incorrectly specified using DEPENDS rather
than DEPENDENCIES

fabaff commented on homeassistant/components/streamlabswater/__init__.py:
```diff
@@ -0,0 +1,112 @@
+
+The away mode service is exposed at streamlabswater.set_away_mode where the
+required parameter away_mode should be set to either home or away.
+"""
```
The content of the docstring should be move to the documentation.

cpopp replied:
Done. Made this consistent with other integrations now that documentation is open for this change.

fabaff commented on homeassistant/components/streamlabswater/__init__.py:
```diff
@@ -0,0 +1,112 @@
+ATTR_AWAY_MODE = 'away_mode'
+SERVICE_SET_AWAY_MODE = 'set_away_mode'
+AWAY_MODE_AWAY = 'away'
+AWAY_MODE_HOME = 'home'
```
Those two constants should be named in a clear way according their content.

cpopp replied:
Do you have an example of what they should be changed to? The modeled the approach after the nest integration for consistency and this is how it declared those constants.

dmulcahey replied:
why not just `HOME = 'home'` and `AWAY = 'away'` as values for something called `DEVICE_MODE`?

Adminiuga replied:
Could there be more modes than `away` or `home`? Why not to make it `is_away` ?

cpopp replied:
The notion of being away seemed like something that could eventually be a general concept so I followed the approach used by climate which calls the service set_away_mode. For consistency and ease of search I used the same constant names used in the nest integration. I don't necessarily disagree with the idea that something different might be more optimal, but until someone designs a common architecture to plug in to I thought being consistent with existing integrations and climate was better.

Adminiuga replied:
I agree to use similar const names as in climate. But in climate "away" is binary. There's no AWAY_MODE_AWAY, AWAY_MODE_HOME etc. I mean "away" notion is binary in its essence. Someone/Something is either away or not. And  climate implements it accordingly, with `turn_away_mode_on()` and `turn_away_mode_off()` and a bool property `is_away_mode_on`

cpopp replied:
@Adminiuga yeah I agree mine looks a bit more like the nest integration than climate in that regard.

fabaff commented on homeassistant/components/streamlabswater/__init__.py:
```diff
@@ -0,0 +1,112 @@
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_AWAY_MODE = 'away_mode'
+SERVICE_SET_AWAY_MODE = 'set_away_mode'
```
This constant is already present.

cpopp replied:
You're referring to the climate constants, right? I opted to not go with those because my device has nothing to do with climate.

fabaff commented on homeassistant/components/streamlabswater/sensor.py:
```diff
@@ -0,0 +1,168 @@
+
+NAME_DAILY_USAGE = "Daily Water"
+NAME_MONTHLY_USAGE = "Monthly Water"
+NAME_YEARLY_USAGE = "Yearly Water"
```
Snake case please

```suggestion
NAME_YEARLY_USAGE = "yearly_water"
```

cpopp replied:
These are just used for the names which as far as I could tell in other sensors are typically capitalized with spaces (or accepted freeform as config parameters).

balloob replied:
Correct. For names spaces and capitalization is allowed. Attribute keys should be snake_case underscore

fabaff commented on homeassistant/components/streamlabswater/sensor.py:
```diff
@@ -0,0 +1,168 @@
+        self._streamlabs_usage_data.update()
+
+
+class StreamLabsMonthlyUsage(Entity):
```
`StreamLabsMonthlyUsage()` and `StreamLabsYearlyUsage()` should be integrated in `StreamLabsDailyUsage`. There is a lot of duplicate code. `StreamLabsUsage` could get the details for a dict/list which describes the different sensors.

cpopp replied:
I changed the monthly and yearly classes to extend the daily usage and that got rid of most of what was common.

fabaff commented on homeassistant/components/streamlabswater/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    locations = client.get_locations().get('locations')
+
+    if locations is None:
+        _LOGGER.error("Unable to retrieve locations. Verify API key.")
```
Please remove the period at the end of all log entries.

fabaff commented on homeassistant/components/streamlabswater/__init__.py:
```diff
@@ -0,0 +1,112 @@
+    for component in STREAMLABSWATER_COMPONENTS:
+        discovery.load_platform(hass, component, DOMAIN, {}, config)
+
+    def set_away_mode(service):
```
Docstring is missing.

fabaff commented on homeassistant/components/streamlabswater/binary_sensor.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up away mode sensor."""
```
```suggestion
    """Set up the StreamLabsWater mode sensor."""
```

cpopp committed:
Streamlabswater pull request feedback

Remove detailed class docstrings since they're in the documentation,
reduce code duplication in sensor classes, and remove periods from
the end of log messages.

cpopp committed:
Reduce line length in Streamlabswater sensor

cpopp committed:
Add docstring on Streamlabswater service callback method

cpopp commented:
I rebased the changes on the latest dev since some conflicting changes were added since I opened the PR.

cpopp committed:
Get rid of unnecessary initializers in Streamlabswater sensor

cpopp committed:
Add manifest file for Streamlabs Water Monitor

fredrike commented on homeassistant/components/streamlabswater/binary_sensor.py:
```diff
@@ -0,0 +1,73 @@
+"""Support for Streamlabs Water Monitor Away Mode."""
```
Why not implement this as a switch and get rid of the `streamlabswater.set_away_mode`service?

cpopp replied:
Two reasons: 1.) When I first started to implement this I considered that but then noticed that would cause it to be controlled along with all other switches in a call to `turn_off` switches when not specifying a particular entity_id (via stuff a user might have set up programatically or via the one click button in the UI to toggle all switches). This seemed pretty risky to me in that disabling away_mode unintentionally could potentially mean a user's home is not being monitored for water user while gone and a leak would go undetected.  So, I thought it was worth making it a separate service so a user could trust it would be operated on independently and integrated in a way.

2.) In general when I surveyed the various switches in home assistant nearly all of them represented turning on or off some physical entity in the real world rather than a conceptual behavior of the entity...so while technically it applicable (it's just a togglable binary value after all) it seemed like it did not fit with the intent of what a switch is.

fredrike replied:
You are absolutely right! But if you build it as an alarm_panel it would make sense, arm away and arm home matches the service call & you can show the triggered state instead of the binary sensor.

cpopp replied:
Right, that makes sense. Long term I think it would be cool for the away mode to integrate with some presence system supported in the core of home assistant so that it gets managed automatically for the user. For now someone could use the [switch template](https://www.home-assistant.io/components/switch.template/) if they wanted it to act like a switch.

fredrike replied:
The "integration with presence system" is done via automations like this one:

```yaml
automation:
  trigger:
  - platform: zone
    entity_id: device_tracker.paulus
    zone: zone.home
    event: leave
  action:
  - service: alarm_control_panel.alarm_arm_away
    data:
      entity_id: alarm_control_panel.streamlabs_water
```

balloob commented on homeassistant/components/streamlabswater/__init__.py:
```diff
@@ -0,0 +1,86 @@
+from homeassistant.helpers import discovery
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['streamlabswater==1.0.1']
```
This constant is no longer used. Please remove.

balloob committed:
Remove unused REQUIREMENTS
